src/v/cluster/archival_metadata_stm.cc

```diff
@@ -620,13 +620,120 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
     co_return co_await builder.replicate();
 }
 
+ss::future<bool>
+archival_metadata_stm::sync(model::timeout_clock::duration timeout) {
+    vlog(_logger.debug, "archival_metadata_stm::sync()");
+    /*
+     * Implementation note
+     * ===================
+     *
+     * This function needs to handle two cases:
+     * Case 1: we have not synced since the last term (e.g. on leadership
+     * changes)
+     * Case 2: we have already synced (once or more) during the current
+     * term, but need to sync again (e.g. re-starting the archiver to apply
+     * topic configs)
+     *
+     * They are handled separately to the reader's delight. In theory, it is
+     * possible to fallthrough from Case 1 to Case 2. This function should be
+     * called in a loop until it succeeds.
+     */
+
+    // If there's any replications pending, wait for them to complete.
+    // Note that we throw away the result and use the Raft's committed offset
+    // below. This deals with both success and failure paths.
+    if (_last_replicate) {
+        auto fut = std::move(_last_replicate.value());
+        _last_replicate = std::nullopt;
+        if (!fut.available()) {
+            vlog(
+              _logger.debug, "Waiting for ongoing replication before syncing");
+
+            const auto before_replication = model::timeout_clock::now();
+
+            try {
+                std::ignore = co_await ss::with_timeout(
+                  before_replication + timeout, std::move(fut));
+            } catch (const ss::timed_out_error&) {
+                vlog(_logger.error, "Replication wait timed out");
+                co_return false;
```

Comment:
I imagine it'd be unlikely, but if we get here, could we run into the same problem? Wouldn't the next iteration of the upload loop sync again, but this time not have a `_last_replicate` to wait on, and potentially proceed without waiting on the still-in-flight replication? On timeout, is there a way to set the future back?

Reply:
It's possible in theory. Evgeny pointed out the same thing. However, in practice I don't think this is a problem.
The timeout that we use is quite long 10s. If we don't replicate within that time we should lose leadership. At that point the term will have changed and we will fall through to `persisted_stm::sync` (old behaviour).

Reply:
> On timeout, is there a way to set the future back?

I think it might be possible to do it with a shared promise, but they're hard to get right.